### PR TITLE
New .NET MAUI Blazor Hybrid template

### LINKED ISSUE #32792: React to new .NET MAUI Blazor Hybrid template
### Description

Per https://github.com/dotnet/AspNetCore.Docs/issues/32692#issuecomment-2155475357.

### Page URL

https://learn.microsoft.com/en-us/aspnet/core/blazor/hybrid/tutorials/maui-blazor-web-app?view=aspnetcore-8.0

### Content source URL

https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md

### Document ID

0aab0118-addc-a59a-0169-0400322d6569

### Article author

@guardrex

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #32792
Addresses #32692
Addresses #31909

Beth ... This should work ... *I think* 🤔🤞🍀. It should work ***IF*** the new project template *exactly matches* what we have in the article in terms of namespace suffixes, file locations and file names, code logic, etc. I still haven't created an app from the template (*YET*) just because I'm working full speed at the moment 🏃‍♂️🏃‍♂️🏃‍♂️ just to get everything covered in time for release tomorrow. I will create a local test app to compare; but if you can tell me about any deltas that exist between the article/sample app and the template, I hope I can work up the changes for this PR and put off creating a local project at the moment.

Here's what this PR is doing to the current coverage ...

* Version a new section to describe the .NET MAUI Blazor Web App project template. BTW ... The workload should already be installed if they follow our earlier guidance here to go work the .NET MAUI Blazor Hybrid tutorial first.
* Version OUT the coverage on *.NET MAUI Blazor Web App sample app* and *Migrating a .NET MAUI Blazor Hybrid solution* so that it only appears for 8.0.
* Keep the rest of the article in place for >=9.0.
* In the *Using interfaces to support different device implementations* section, refactor the language to describe the API in a way that should cover both 8.0 folks using it to manually build the app and 9.0 or later folks who are getting it generated from the project template ... as I mentioned, ***IF*** all of the naming and code was preserved exactly for the new template.

WRT the command, I included the interactivity and interactivity location options, but you might prefer that I strike them because they're probably the default ...

```dotnetcli
dotnet new maui-blazor-web -o MauiBlazorWeb -int Server -ai
```

If you confirm that they are the default, I'll change the text a bit and remove them from that command. I'll tell the reader the defaults in the text next to the command.

I'll retain the option for naming the folder/namespace (`-o|--output MauiBlazorWeb`).

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md](https://github.com/dotnet/AspNetCore.Docs/blob/a377cd5a754f5257f7e844a44503c13a54977c6f/aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md) | [Build a .NET MAUI Blazor Hybrid app with a Blazor Web App](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/hybrid/tutorials/maui-blazor-web-app?branch=pr-en-us-32793) |

### CODE REVIEW

File: aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md [hunk 1]
```diff
@@ -16,6 +16,31 @@ This article shows you how to build a .NET MAUI Blazor Hybrid app with a Blazor
 
 For prerequisites and preliminary steps, see <xref:blazor/hybrid/tutorials/maui>. We recommend using the .NET MAUI Blazor Hybrid tutorial to set up your local system for .NET MAUI development before using the guidance in this article.
 
+:::moniker range=">= aspnetcore-9.0"
+
+## .NET MAUI Blazor Web App project template
```
Comment (BethMassi) [inline aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md:21]:
```suggestion
## .NET MAUI Blazor Hybrid and Web App solution template
```

File: aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md [hunk 2]
```diff
@@ -16,6 +16,31 @@ This article shows you how to build a .NET MAUI Blazor Hybrid app with a Blazor
...
+
+The .NET MAUI Blazor Web App project template allows you to choose a Blazor interactive render mode for the web app and creates the appropriate projects for the app, including a Blazor Web App and a .NET MAUI Blazor Hybrid app. A shared Razor class library (RCL) maintains the Razor components for the app's UI. The template also provides sample code to shows you how to use dependency injection to provide different interface implementations for the Blazor Hybrid and Blazor Web App, which is covered in the [Using interfaces to support different device implementations](#using-interfaces-to-support-different-device-implementations) section of this article.
```
Comment (BethMassi) [inline aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md:23]:
```suggestion
The .NET MAUI Blazor Hybrid and Web App solution template sets up a solution that targets Android, iOS, Mac, Windows and Web that reuses UI. You can choose a Blazor interactive render mode for the web app and it creates the appropriate projects for the app, including a Blazor Web App and a .NET MAUI Blazor Hybrid app. A shared Razor class library (RCL) maintains the Razor components for the app's UI. The template also provides sample code to show you how to use dependency injection to provide different interface implementations for the Blazor Hybrid and Blazor Web App, which is covered in the [Using interfaces to support different device implementations](#using-interfaces-to-support-different-device-implementations) section of this article.
```

File: aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md [hunk 3]
```diff
@@ -16,6 +16,31 @@ This article shows you how to build a .NET MAUI Blazor Hybrid app with a Blazor
...
+* The `-ai|--all-interactive` option specifies global interactivity, which is important because MAUI apps always run interactively and throw errors on Razor component pages that explicitly specify a render mode. If you don't use a global render mode, you must implement the approach described in the [Use Blazor render modes](#use-blazor-render-modes) section. For more information, see [BlazorWebView needs a way to enable overriding ResolveComponentForRenderMode (`dotnet/aspnetcore` #51235)](https://github.com/dotnet/aspnetcore/issues/51235).
+
+<!-- UPDATE 9.0 Provide the project template's name here for VS, possibly
+                using a tooling pivot for the article. -->
+
+:::moniker-end
+
+:::moniker range="< aspnetcore-9.0"
+
 ## .NET MAUI Blazor Web App sample app
```
Comment (BethMassi) [inline aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md:44]:
```suggestion
## .NET MAUI Blazor Hybrid and Web App sample app
```

File: aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md [hunk 4]
```diff
@@ -365,19 +392,19 @@ In the `_Imports.razor` file of the `.Shared.Client` RCL, add `@using static Int
...
 
-In the RCL's `Components` folder, add the following `DeviceFormFactor` component.
+The following `DeviceFormFactor` component is present in the RCL's `Components` folder:
 
 `Components/Pages/DeviceFormFactor.razor`:
 
 :::code language="razor" source="~/../blazor-samples/8.0/MauiBlazorWeb/MauiBlazorWeb.Shared/Components/Pages/DeviceFormFactor.razor":::
 
-In the RCL, add an entry for the `DeviceFormFactor` component to the navigation menu.
+In the RCL, an entry for the `DeviceFormFactor` component is part of the navigation menu in the `NavMenu` component.
```
Comment (BethMassi) [inline aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md:407]: In the new solution template, the form factor UI is on the home page. There isn't a DeviceFormFactor.razor page and there is no change to the NavMenu.

File: aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md (hunk 3 above)
Comment (BethMassi) [reply aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md:44]: The only difference between the guidance below and what the solution template creates for you is that the name of the MAUI project does not have ".Maui" naming, it's just the name that the user indicated as the name of their project/solution. So if you named it "MyApp" it would be: 
  * Solution: `MyApp`
  * MAUI project: `MyApp`
  * Blazor Web App: `MyApp.Web`
  * Blazor Web Assembly App: `MyApp.Web.Client`
  * Razor class library (RCL): `MyApp.Shared`

File: aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md [hunk 5]
```diff
@@ -389,43 +416,43 @@ In `Components/Layout/NavMenu.razor`:
...
 :::code language="csharp" source="~/../blazor-samples/8.0/MauiBlazorWeb/MauiBlazorWeb.Maui/Services/FormFactor.cs":::
 
-Use dependency injection to obtain the implementations of these services.
+Dependency injection is used to obtain the implementations of these services.
 
-In the MAUI project, open the `MauiProgram.cs` file and add the following `using` statements to the top of the file:
+In the MAUI project, the `MauiProgram.cs` file has following `using` statements at the top of the file:
 
 ```csharp
 using MauiBlazorWeb.Maui.Services;
```
Comment (BethMassi) [inline aspnetcore/blazor/hybrid/tutorials/maui-blazor-web-app.md:438]:
```suggestion
using MauiBlazorWeb.Services;
```

**Review by BethMassi — CHANGES_REQUESTED** [review]
Thanks! Some minor changes needed.

**Comment by guardrex** [comment]
@BethMassi ... This will be a *Texas Two-step* operation. First, I'll merge your commits, then I'll see about a commit to deal with the `.Maui` naming and how the device form factor piece moved to the home page. I'll try to version it; otherwise, it will break the current sample app. Perhaps, I can find time later to make the 8.0 sample app conform to the new template and we can drop the versioning, but I don't think that can be done in the very near-term due to the workload right now.

**Comment by guardrex** [comment]
@BethMassi ... See if my last commit resolves ...

* Naming ... drop `.Maui` from project names for 9.0 or later, and keep it for 8.0.
* Form factor and platform ... keep existing `DeviceFormFactor` component coverage for 8.0, and show the new `Home` component for 9.0 or later.

**Comment by guardrex** [comment]
... and BTW ... I will open a follow-up issue to make the 8.0 sample app match that. Then, I can go back in here later and nix the versioned content and use the new naming.

**Comment by guardrex** [comment]
Also, there was some kind of build failure on the prior commit. I'll check on that after this builds and see what broke it 💥.

**Comment by guardrex** [comment]
Ok ... I see the problem. I'll fix it.

**Comment by guardrex** [comment]
# 🎉 

Success! See how it looks now with the latest updates.

**Review by BethMassi — APPROVED** [review]
I still see references to MauiBlazorWeb.Maui but as long as that is only for 8.0 that's OK I guess? It's hard to follow the conditional statements :-) so I trust you. I'd suggest just updating the .NET 8 sample to match the naming so it's easier.

**Comment by guardrex** [comment]
>  It's hard to follow the conditional statements :-)

***Exactly!*** ... which is why I've been arguing for **Version-by-File** for years. They won't implement it (yet anyway), so we're stuck with this versioning for the time being.

Yes, I searched the doc and updated the versioning for the `.Maui` naming, and I'll give it a final look before merging this.

> I'd suggest just updating the .NET 8 sample to match the naming so it's easier.

It will take longer than doing it this way. I'll fix that up on the other issue that I opened. Hopefully, I'll reach that in the RC time frame.

**Comment by guardrex** [comment]
***ORF!*** My power is going ***MAD*** here right now. We're getting slammed by a giant thunderstorm :zap:, and I'm losing power and Internet intermittently. I'll try to wrap this up ***between blackouts!*** 😆

**Comment by guardrex** [comment]
Ok ... we're good. I'll merge now.
